packages/protocol/contracts/stability/BreakerBox.sol

```diff
@@ -0,0 +1,305 @@
+
+    uint256 lastIndex = exchanges.length - 1;
+    if (exchangeIndex != lastIndex) {
+      exchanges[exchangeIndex] = exchanges[lastIndex];
+    }
+
+    exchanges.pop();
+
+    delete exchangeTradingModes[exchange];
+    emit ExchangeRemoved(exchange);
+  }
+
+  /**
+   * @notice Sets the trading mode for the specified exchange.
+   * @param exchange The address of the exchange.
+   * @param tradingMode The trading mode that should be set.
+   */
+  function setExchangeTradingMode(address exchange, uint64 tradingMode) public onlyOwner {
+    require(
+      tradingMode == 0 || tradingModeBreaker[tradingMode] != address(0),
+      "Trading mode must be default or have a breaker set"
+    );
+
+    TradingModeInfo memory info = exchangeTradingModes[exchange];
+    require(info.lastUpdatedTime > 0, "Exchange has not been added");
+
+    info.tradingMode = tradingMode;
+    info.lastUpdatedTime = uint64(block.timestamp);
+    info.lastUpdatedBlock = uint128(block.number);
+    exchangeTradingModes[exchange] = info;
+
+    emit TradingModeUpdated(exchange, tradingMode);
+  }
+
+  /* ==================== View Functions ==================== */
+
+  /**
+   * @notice Returns an array of breaker addresses from start to end.
+   * @return An ordered list of breakers.
+   */
+  function getBreakers() external view returns (address[] memory) {
+    return breakers.getKeys();
+  }
+
+  /**
+   * @notice Checks whether a breaker with the specifed address has been added.
+   */
+  function isBreaker(address breaker) public view returns (bool) {
+    return breakers.contains(breaker);
+  }
+
+  /**
+   * @notice Returns addresses of exchanges that have been added.
+   */
+  function getExchanges() external view returns (address[] memory) {
+    return exchanges;
+  }
+
+  /* ==================== Check Breakers ==================== */
+
+  /**
+   * @notice Checks breakers for a specified exchange to determine the trading mode.
+   * @param exchangeAddress The address of the exchange to run the checks for.
+   * @return currentTradingMode Returns an int representing the current trading mode
+   *                            for the specified exchange.
+   */
+  function checkBreakers(address exchangeAddress) external returns (uint256 currentTradingMode) {
+    TradingModeInfo memory info = exchangeTradingModes[exchangeAddress];
+
+    // Last updated should always have a value gt 0
+    // if it doesn't we can assume this exchange has not been added.
+    if (info.lastUpdatedTime == 0) {
+      return 0;
+    }
+
+    // Check if a breaker has non default trading mode & try to reset
+    if (info.tradingMode != 0) {
+      IBreaker breaker = IBreaker(tradingModeBreaker[info.tradingMode]);
+
+      uint256 cooldown = breaker.getCooldown();
+      bool tryReset = ((cooldown > 0) && (cooldown + info.lastUpdatedTime) <= block.timestamp);
+
+      if (tryReset) {
+        bool canReset = breaker.shouldReset(exchangeAddress);
+        if (canReset) {
```

Comment:
Any reason to separate this into two lines instead of performing the check in the `if`?

Reply:
Another thing to consider: are the `ResetAttemptCriteriaFail` and `ResetAttemptNotCool` particularly useful? If not, we could simplify this to a single `if` instead of nesting.

Reply:
> Any reason to separate this into two lines instead of performing the check in the `if`?
Nope, they can be inlined.

> Another thing to consider: are the `ResetAttemptCriteriaFail` and `ResetAttemptNotCool` particularly useful? If not, we could simplify this to a single `if` instead of nesting.

They'll be useful for monitoring. If a breaker is tripped these logs will let us easily determine why the trading mode has not been reset yet. Also at some point in the future, I'd like to put together a UI or a dashboard to show this info.